### PR TITLE
fix: error filtering of non-errors

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

- Fixed error filtering for non-errors
- Enable error reporting

## Linear task (optional)

[Linear task link](https://linear.app/rudderstack/issue/SDK-2176/fix-error-reporting-plugin-issue)

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
	- Enhanced error reporting functionality, allowing dynamic enabling or disabling based on configuration.
	- Introduced a function to improve detection of non-error messages in error reporting logic.

- **Bug Fixes**
	- Improved the overall control flow in error handling to reduce the risk of runtime errors.
	- Simplified error handling logic to enhance clarity and performance.

- **Tests**
	- Added new test cases to improve coverage for error handling utilities.
	- Reactivated a previously skipped test case to validate error handling logic related to SDK script targets.

### CODE REVIEW

File: packages/analytics-js-plugins/src/errorReporting/utils.ts [hunk 1]
```diff
@@ -144,6 +144,10 @@ const getBugsnagErrorEvent = (
 
 const isRudderSDKError = (event: any) => {
   const errorOrigin = event.stacktrace?.[0]?.file;
+  const errorMessage = event.message;
+  if (errorMessage && typeof errorMessage === 'string' && errorMessage.includes('a non-error')) {
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/utils.ts:148]: The string "a non-error" doesn't look reliable enough. Do we have anything else?

Comment (saikumarrs) [reply packages/analytics-js-plugins/src/errorReporting/utils.ts:148]: Why can't we use the error message filtering in the core SDK that we already have? We shouldn't filter errors this way. If needed, use regex based string matching.

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 2]
```diff
@@ -73,14 +73,10 @@ const normaliseError = (
   let error;
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/event/event.ts:73]: We don't need any complex logic to handle different types of the errors. Also, remove tolerateNonErrors input.

if (error) {
  // proceed
} else {
  // log warning
  // proceed
}

// throw error if needed.

File: packages/analytics-js-plugins/src/errorReporting/utils.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ import type {
 import { generateUUID } from '@rudderstack/analytics-js-common/utilities/uuId';
 import { METRICS_PAYLOAD_VERSION } from '@rudderstack/analytics-js-common/constants/metrics';
 import { stringifyWithoutCircular } from '@rudderstack/analytics-js-common/utilities/json';
+import { ERROR_MESSAGES_TO_BE_FILTERED } from '@rudderstack/analytics-js-common/constants/errors';
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/utils.ts:20]: It doesn't make sense to be part of the common package anymore.

File: packages/analytics-js-plugins/src/errorReporting/utils.ts [hunk 4]
```diff
@@ -142,8 +143,25 @@ const getBugsnagErrorEvent = (
...
+
+/**
+ * A function to determine if the error is from Rudder SDK
+ * @param {Error} event
+ * @returns
+ */
 const isRudderSDKError = (event: any) => {
   const errorOrigin = event.stacktrace?.[0]?.file;
+  const errorMessage = event.message;
+  if (errorMessage && typeof errorMessage === 'string' && !isAllowedToBeNotified(errorMessage)) {
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/utils.ts:162]: Do not combine error filtering action into this method. It has a different responsibility.

File: packages/analytics-js-common/src/constants/errors.ts
Comment (saikumarrs) [inline packages/analytics-js-common/src/constants/errors.ts:1]: You need to remove this constant `UNHANDLEDEXCEPTION_FOR_NON_ERROR_OBJECT` as it is not required anymore.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: packages/analytics-js/src/services/HttpClient/xhr/xhrRequestHandler.ts [hunk 5]
```diff
@@ -8,8 +8,8 @@ import type {
 } from '@rudderstack/analytics-js-common/types/HttpClient';
 import type { ILogger } from '@rudderstack/analytics-js-common/types/Logger';
 import { getMutatedError } from '@rudderstack/analytics-js-common/utilities/errors';
+import { FAILED_REQUEST_ERR_MSG_PREFIX } from '@rudderstack/analytics-js-common/constants/errors';
```
Comment (saikumarrs) [inline packages/analytics-js/src/services/HttpClient/xhr/xhrRequestHandler.ts:11]: Does it make sense to move this constant into this module?

File: packages/analytics-js-plugins/src/errorReporting/utils.ts (hunk 1 above)
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/utils.ts]: Addressed

File: packages/analytics-js-plugins/src/errorReporting/utils.ts (hunk 3 above)
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/utils.ts:20]: Please resolve it

File: packages/analytics-js-common/src/constants/errors.ts
Comment (MoumitaM) [reply packages/analytics-js-common/src/constants/errors.ts:1]: Addressed

File: packages/analytics-js/src/services/HttpClient/xhr/xhrRequestHandler.ts (hunk 5 above)
Comment (MoumitaM) [reply packages/analytics-js/src/services/HttpClient/xhr/xhrRequestHandler.ts:11]: As per our discussion please resolve it

File: packages/analytics-js-plugins/src/errorReporting/index.ts [hunk 6]
```diff
@@ -73,6 +74,9 @@ const ErrorReporting = (): ExtensionPlugin => ({
         const { component, tolerateNonErrors, errorFramesToSkip, normalizedError } =
           getConfigForPayloadCreation(error, errorState?.severityReason.type as string);
 
+        if (!isAllowedToBeNotified(error)) {
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/index.ts:77]: Why don't we perform this check after creating the payload? By then, the error message would also have been apparent.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 7]
```diff
@@ -64,71 +64,22 @@ const hasNecessaryFields = (error: any) =>
...
-  if (!tolerateNonErrors) {
-    if (isError(maybeError)) {
-      error = maybeError;
-    } else {
-      error = createAndLogInputError(typeof maybeError);
-      internalFrames += 2;
-    }
+  if (maybeError !== null && isError(maybeError)) {
+    error = maybeError;
+  } else if (maybeError !== null && hasNecessaryFields(maybeError)) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts:73]: Yes that checking is done at Error reporting module

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 8]
```diff
@@ -137,7 +88,7 @@ const normaliseError = (
         error = e;
         // if the error only got a stacktrace after we threw it here, we know it
         // will only have one extra internal frame from this function, regardless
-        // of whether it went through createAndLogInputError() or not
+        // of whether it went through logInputError() or not
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Done

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 9]
```diff
@@ -64,71 +64,22 @@ const hasNecessaryFields = (error: any) =>
...
-        } else {
-          error = createAndLogInputError(maybeError === null ? 'null' : 'unsupported object');
-          internalFrames += 2;
-        }
-        break;
-      default:
-        error = createAndLogInputError('nothing');
-        internalFrames += 2;
-    }
+    if (logger) logger.warn(error);
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Done

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Done

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 10]
```diff
@@ -64,71 +64,22 @@ const hasNecessaryFields = (error: any) =>
...
-  // tolerateNonErrors option to ensure that the resulting error communicates as
-  // such.
-  if (!tolerateNonErrors) {
-    if (isError(maybeError)) {
-      error = maybeError;
-    } else {
-      error = createAndLogInputError(typeof maybeError);
-      internalFrames += 2;
-    }
+  if (maybeError !== null && isError(maybeError)) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Removed it

File: packages/analytics-js-plugins/src/errorReporting/index.ts (hunk 6 above)
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/index.ts]: Done

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 11]
```diff
@@ -64,80 +66,33 @@ const hasNecessaryFields = (error: any) =>
...
-        }
-        break;
-      default:
-        error = createAndLogInputError('nothing');
-        internalFrames += 2;
-    }
+    if (logger)
+      logger.warn(
+        `${ERROR_REPORTING_PLUGIN}:: ${component} received a non-error: ${stringifyWithoutCircular(error)}`,
+      );
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/event/event.ts:83]:
```suggestion
    logger?.warn(
        `${ERROR_REPORTING_PLUGIN}:: ${component} received a non-error: ${stringifyWithoutCircular(error)}`,
      );
```

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts [hunk 12]
```diff
@@ -64,80 +66,33 @@ const hasNecessaryFields = (error: any) =>
...
-  // tolerateNonErrors option to ensure that the resulting error communicates as
-  // such.
-  if (!tolerateNonErrors) {
-    if (isError(maybeError)) {
-      error = maybeError;
-    } else {
-      error = createAndLogInputError(typeof maybeError);
-      internalFrames += 2;
-    }
+  if (typeof maybeError === 'object' && isError(maybeError)) {
```
Comment (saikumarrs) [inline packages/analytics-js-plugins/src/errorReporting/event/event.ts:73]: I did not suggest the object check here but for the next else if.

```suggestion
  if (isError(maybeError)) {
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]
LGTM except for these minor updates.

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts (hunk 11 above)
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Addressed

File: packages/analytics-js-plugins/src/errorReporting/event/event.ts (hunk 12 above)
Comment (MoumitaM) [reply packages/analytics-js-plugins/src/errorReporting/event/event.ts]: Addressed

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**